library/src/main/java/com/mux/player/media/MuxDataSource.kt

```diff
@@ -98,4 +101,209 @@ class MuxDataSource private constructor(
+
+      openAndInitFromCache(readHandle)
+    }
+  }
+
+  private fun openAndInitFromRemote(dataSpec: DataSpec, fac: HttpDataSource.Factory): Long {
+    respondingFromCache = false
+    val upstream = fac.createDataSource()
+    this.upstream = upstream
+    val available = upstream.open(dataSpec)
+    cacheWriter = CacheController.downloadStarted(
+      dataSpec.uri.toString(),
+      upstream.responseHeaders,
+    )
+    return available
+  }
+
+  private fun openAndInitFromCache(readHandle: ReadHandle): Long {
+    respondingFromCache = true
+    this.cacheReader = readHandle
+    return readHandle.fileSize
+  }
+}
+
+/**
+ * HttpDataSource used for revalidating stale segments.
+ *
+ * This is required because the DefaultHttpDataSource doesn't support 304 responses
+ * It's ok for this class to be simple because the main special cases covered by the
+ * DefaultHttpDataSource are not relevant for this version of the cache. Chunks aren't gzipped,
+ * never requested in parts, and mux video doesn't redirect segment requests so we can just
+ * straightforwardly read the bytes
+ */
+@OptIn(UnstableApi::class)
+private class RevalidatingDataSource : BaseDataSource(true), HttpDataSource {
+
+  class Factory: HttpDataSource.Factory {
+    override fun createDataSource(): HttpDataSource {
+      return RevalidatingDataSource()
+    }
+    override fun setDefaultRequestProperties(defaultRequestProperties: MutableMap<String, String>): HttpDataSource.Factory {
+      // not used
+      return this
+    }
+  }
+
+  private var httpConnection: HttpURLConnection? = null
+  private var bodyInputSteam: InputStream? = null
+  private var responseCode: Int? = null
+  private var responseMessage: String? = null
+
+  private var open: Boolean = false
+  private var dataSpec: DataSpec? = null
+  private val requestHeaders = mutableMapOf<String, String>()
+
+  override fun read(buffer: ByteArray, offset: Int, length: Int): Int {
+    val readBytes = bodyInputSteam?.read(buffer, offset, length) ?: 0
+
+    return if (readBytes == -1) {
+      C.RESULT_END_OF_INPUT
+    } else {
+      bytesTransferred(readBytes)
+      readBytes
+    }
+  }
+
+  override fun open(dataSpec: DataSpec): Long {
+    val conn = try {
+      val hurlConn = createConnection(dataSpec, requestHeaders)
+      this.responseCode = hurlConn.responseCode
+      this.responseMessage = hurlConn.responseMessage
+      hurlConn
+    } catch (e: IOException) {
+      closeConnection()
+      throw HttpDataSourceException.createForIOException(
+        e, dataSpec,
+        HttpDataSourceException.TYPE_OPEN
+      )
+    }
+
+    val code = conn.responseCode
+    val msg = conn.responseMessage
+    if (code == HttpURLConnection.HTTP_NOT_MODIFIED) {
+      // not-modified, not an error, we don't have to download the body again
+      return 0
+    } else if (code < 200 || code > 299) {
+      // some kind of error
+      val errorBody = conn.errorStream.use { errorBody -> errorBody.readBytes() }
```

Comment:
question(blocking): I don't **think** we need to account for 3xx range statuses like redirects (for Mux Video), but we should get an answer to this before merging.

Reply:
Just checked in with Mike Smith. This should be good enough for now! 🎉